R/repl.R

```diff
@@ -0,0 +1,171 @@
+#' Run a Python REPL
+#' 
+#' This function provides a Python REPL in the R session, which can be used
+#' to interactively run Python code. All code executed within the REPL is
+#' run within the Python 'main' module, and will remain accessible after the
+#' REPL is detached.
+#' 
+#' @param module An (optional) Python module to be imported before
+#'   the REPL is launched. Defaults to the value of the `reticulate.repl.module`
+#'   option.
+#' @param quiet Boolean; print a startup banner when launching the REPL? If
+#'   `FALSE`, the banner will be suppressed.
+#' 
+#' @export
+py_repl <- function(
+  module = getOption("reticulate.repl.module", default = character()),
+  quiet = getOption("reticulate.repl.quiet", default = FALSE))
+{
+  # load modules (do this first so that we bind to the expected version of
+  # Python if this is the first attempt to load Python)
+  for (m in module)
+    import(m)
+  
+  # import other required modules for the REPL
+  builtins <- import_builtins(convert = FALSE)
+  main <- import_main(convert = FALSE)
+  codeop <- import("codeop", convert = TRUE)
+  
+  # grab references to the locals, globals of the main module
+  locals <- py_run_string("locals()")
+  globals <- py_run_string("globals()")
+  
+  # check to see if the current environment supports history
+  has_history <- tryCatch(
+    { utils::savehistory(tempfile()); TRUE },
+    error = function(e) FALSE
+  )
+  
+  if (has_history) {
```

Comment:
Does this mechanism also account for the RStudio global history? (IIRC we don't save entires from the non-default prompt but I'm not sure if that's the case)

Reply:
It does! Effectively, there are two separate histories -- one for R, and one for Python, and the active history gets toggled on / off as the REPL is activated / deactivated.